bql/semantic/hooks_test.go

```diff
@@ -179,8 +179,220 @@ func TestWhereInitClauseHook(t *testing.T) {
 	f := whereInitWorkingClause()
 	st := &Statement{}
 	f(st, Symbol("FOO"))
-	if st.WorkingClause() == nil {
-		t.Errorf("semantic.WhereInitWorkingClause should have returned a valid working clause for statement %v", st)
+	if wc := st.WorkingClause(); wc == nil || !wc.IsEmpty() {
+		t.Errorf(`semantic.Statement.WorkingClause() = %q for statement "%v" after call to semantic.WhereInitWorkingClause; want empty GraphClause`, wc, st)
+	}
+	if wf := st.WorkingFilter(); wf == nil || !wf.IsEmpty() {
+		t.Errorf(`semantic.Statement.WorkingFilter() = %q for statement "%v" after call to semantic.WhereInitWorkingClause; want empty FilterClause`, wf, st)
+	}
+}
+
+func TestWhereFilterClauseHook(t *testing.T) {
+	st := &Statement{}
+	f := whereFilterClause()
+	st.ResetWorkingFilterClause()
+
+	testTable := []struct {
+		id   string
+		ces  []ConsumedElement
+		want *FilterClause
+	}{
+		{
+			id: "FILTER latest(?p)",
+			ces: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemFilter,
+					Text: "FILTER",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemFilterFunction,
+					Text: "latest",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemLPar,
+					Text: "(",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?p",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemRPar,
+					Text: ")",
+				}),
+				NewConsumedSymbol("FOO"),
+			},
+			want: &FilterClause{
+				Operation: "latest",
+				Binding:   "?p",
+			},
+		},
+		{
+			id: "FILTER latest(?o)",
+			ces: []ConsumedElement{
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemFilter,
+					Text: "FILTER",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemFilterFunction,
+					Text: "latest",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemLPar,
+					Text: "(",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemBinding,
+					Text: "?o",
+				}),
+				NewConsumedToken(&lexer.Token{
+					Type: lexer.ItemRPar,
+					Text: ")",
+				}),
+				NewConsumedSymbol("FOO"),
+			},
+			want: &FilterClause{
+				Operation: "latest",
+				Binding:   "?o",
+			},
+		},
+	}
+	for i, entry := range testTable {
+		for _, ce := range entry.ces {
+			if _, err := f(st, ce); err != nil {
+				t.Errorf("%q: semantic.WhereFilterClauseHook(%s) = _, %v; want _, nil", entry.id, ce, err)
+			}
+		}
+		if got, want := st.FilterClauses()[i], entry.want; !reflect.DeepEqual(got, want) {
+			t.Errorf("%q: semantic.Statement.FilterClauses()[%d] = %s; want %s", entry.id, i, got, want)
+		}
+	}
```

Comment:
I recommend using t.Run to run subtests one for each table entry. Reference:https://gobyexample.com/testing.

You can always introduce a "description" field to your table entries, and pass it as the first parameter of t.Run. This way the code is self documented because each test case has a descriptive string. e.g
       desc: "FilterClauseHook returns error"
        id: "FILTER latest(?p)",
       ces: []ConsumedElement{}

Reply:
Done. Thank you Jose.

Here I used the `entry.id` as the test description since, given the name of the function, we already know that we are testing for success. For `TestWhereFilterClauseHookError` below I do the same thing since, given the name of the function, we already know that we are testing for error.

Reply:
Related Issue: https://github.com/google/badwolf/issues/152.